Generate and upload resource cucumber tests

jeddcuison opened the pull request:
This is to verify the accuracy of the merge of master and cucumber-rahul since the later branch is way behind due to infrequent pull and merge from master

jeddcuison committed:
Initial Cucumber tests

Details:

- Use of DI via Spring
- Establish run step patterns
- Do cleanup after running each tests
- Ignore Test Suite (this will be deprecated soon)

jeddcuison committed:
Add README.md

Instructions on how to run tests

jeddcuison committed:
Delete group jvm table to prevent foreign key constraint error when deleting the group and the jvm data

jeddcuison committed:
Test media creation

jeddcuison committed:
Add file.upload.dir property

jeddcuison committed:
Rename method

jeddcuison committed:
Test web server creation

jeddcuison committed:
Test web app creation

jeddcuison committed:
Add main ui test and exclude all other tests from running with Gradle test

jeddcuison committed:
Delete to fix filename upper/lower case problem

jeddcuison committed:
Add test with the correct filename camel case

jeddcuison committed:
Click the tab when busy screen is not displayed anymore

jeddcuison committed:
Do cleanup after each scenario

jeddcuison committed:
Delete Test class

jeddcuison committed:
Change hard coded path

jeddcuison committed:
Configure test to generate Cucumber report in json

jeddcuison committed:
Remove deprecated Cucumber test

rahulsayini committed:
Merge branch 'cucumber' into cucumber-rahul

cucumber and selenium tests

rahulsayini committed:
Add webserver cucumber test cases

rahulsayini committed:
remove sleep method

rahulsayini committed:
add more cucumber test cases for resource upload

rahulsayini committed:
update test.properties file

jeddcuison committed:
Test JVM creation

rahulsayini committed:
rename feature file

rahulsayini committed:
Merge branch 'cucumber' into cucumber-rahul

# Conflicts:
#	jwala-integration-test/src/test/java/com/cerner/jwala/ui/selenium/steps/configuration/ManageJvmRunSteps.java

jeddcuison committed:
Check for web app name, context and group

jeddcuison committed:
Add waitUntilElementIstVisible method with timeout parameter

rahulsayini committed:
add more selenium test cases for resource upload

rahulsayini committed:
Merge branch 'cucumber' into cucumber-rahul

rahulsayini committed:
add cucumber test cases for upload resources

rahulsayini committed:
Add OperationRunSteps.java

rahulsayini committed:
add metadata check option for resource upload

rahulsayini committed:
Merge branch 'master' into cucumber-rahul

# Conflicts:
#	jwala-integration-test/src/test/java/com/cerner/jwala/ui/selenium/component/JwalaUi.java
#	jwala-integration-test/src/test/java/com/cerner/jwala/ui/selenium/steps/CommonRunSteps.java

rahulsayini committed:
add more test cases for generate and deploy

rahulsayini committed:
fill status path and avoid auto-fill of status path

rahulsayini committed:
address code review comments

rahulsayini committed:
Address code review comments

rahulsayini committed:
add "When" for cucumber scenario best practices and follow Given-When-Then

Sharvaribarve committed:
Added properties

Sharvaribarve committed:
Added host

jeddcuison committed:
Refactor

jeddcuison committed:
Resolve merge conflict

jeddcuison committed:
Remove duplicate method

jeddcuison committed:
Use spring to inject the parameters properties object

jeddcuison committed:
Use the method with the corrected method name

jeddcuison committed:
Change resourcesDeployPath to resource.deploy.path

jeddcuison committed:
Remove getJwalaProperties

jeddcuison committed:
Add line

jeddcuison committed:
Refactor upload resource integration test

- consolidate JVM, web server and web application resource upload scenarios into one feature
- remove "I load properties file"
- maintain only one unit test to run JVM, web server and web application resource upload

Sharvaribarve commented on jwala-integration-test/src/test/java/com/cerner/jwala/ui/selenium/steps/configuration/UploadResourceRunSteps.java:
```diff
@@ -0,0 +1,84 @@
+
+    @When("^check resource uploaded successful")
+    public void checkForResource() {
+        jwalaUi.waitUntilElementIsVisible(By.xpath("//input[contains(@class, 'noSelect')]/following-sibling::span"));
```
Could change it to Then rather than when, and we could specify the resource name, to check to see if it is succcesfully uploaded.

jeddcuison committed:
Make the feature more readable and tweak the parameter properties file

jeddcuison committed:
Ignore the feature for now since it needs more work

Other change(s)

- Change remoteDir to media.remote.dir

jeddcuison committed:
Define tag @ignore

jeddcuison committed:
Ignore the feature for now since it needs more work

jeddcuison committed:
Refactor

- make the scenarios more readable
- add expand and collapse node methods

jeddcuison committed:
Do scenario cleanup only once

jeddcuison committed:
Remove "Manage" prefix

jeddcuison committed:
Refactor

- make xpaths for concise since some xpaths afre generic and can lead to unexpected testing errors
- make the scenarios more readable
- organize step methods better

jeddcuison committed:
Remove old tests

jeddcuison committed:
Fix bug wherein the browser driver is still running after the tests are ran

jeddcuison committed:
Remove feature file that was not supposed to be committed

The said file was a sandbox file meant to show something related to properties to another developer but was committed by accident

jeddcuison committed:
Include jwala-integration-test

In addition:

- Exclude running of jwala-integration-test's via test task

jeddcuison committed:
Update java doc and remote unused constants

jeddcuison committed:
Remove unused class

jeddcuison committed:
Refactor generate and deploy feature and remove created windows services on cleanup

 On Refactor:

 - reword the steps to make it more concise and readable
 - fix bug wherein Cucumber does not run some of the steps by redefining the steps (e.g. instead of one "operational" step, we write smaller ones e.g. start and stop steps)

jeddcuison committed:
Define ssh user and pwd

jeddcuison committed:
Remove commented out code

Sharvaribarve commented on jwala-integration-test/src/test/java/com/cerner/jwala/ui/selenium/steps/operation/GenerateJvmRunSteps.java:
```diff
@@ -0,0 +1,28 @@
+    public void checkForTheSuccessfulGenerationOfJvms(final String groupName) {
+        jwalaUi.waitUntilElementIsVisible(By.xpath("//div[text()='Successfully generated the JVMs for " + groupName + "']"), 600);
+        jwalaUi.click(By.xpath("//button[span[text()='Ok']]"));
+        jwalaUi.sleep();
```
Why do we need sleep after the success message?

jeddcuison replied:
No. I still have to cleanup this branch before I put the story in the review column. This will be removed.

Sharvaribarve commented on jwala-integration-test/src/test/java/com/cerner/jwala/ui/selenium/steps/operation/GenerateWebServerRunSteps.java:
```diff
@@ -0,0 +1,27 @@
+
+    @When("^I click the \"Generate Web Servers\" button of group \"(.*)\"$")
+    public void clickGenerateWebServersBtnOfGroup(final String groupName) {
+        jwalaUi.click(By.xpath("//tr[td[text()='" + groupName + "']]/following-sibling::tr//button[span[text()='Generate Web Servers']]"));
```
I generally prefer using clickWhenReady instead in operations page since the  gears are visible sometime after expanding the group

Sharvaribarve commented on jwala-integration-test/src/test/java/com/cerner/jwala/ui/selenium/steps/operation/GenerateWebAppRunSteps.java:
```diff
@@ -0,0 +1,27 @@
+
+    @When("^I click the generate web application button of \"(.*)\"$")
+    public void clickGenerateWebAppBtnOfGroup(final String webAppName) {
+        jwalaUi.click(By.xpath("//tr/td[text()='" + webAppName + "']/following-sibling::td[3]/div/button[@title='Generate and deploy the webapp resources.']"));
```
I generally prefer using clickWhenReady instead in operations page since the gears are visible sometime after expanding the group

jeddcuison replied:
Nice catch

Sharvaribarve replied:
Thanks I will check mine :)

Sharvaribarve commented on jwala-integration-test/src/test/java/com/cerner/jwala/ui/selenium/steps/configuration/UploadResourceRunSteps.java:
```diff
@@ -0,0 +1,84 @@
+    private Properties paramProp;
+
+    @Given("^I am in the resource tab$")
+    public void goToConfigurationTab() {
```
Method name needs to be changed

jeddcuison replied:
Nice catch, I just renamed this file from [Rahul's](https://github.com/cerner/jwala/blob/cae1c795a1c2d9a0927f62da8357706a562e4e7a/jwala-integration-test/src/test/java/com/cerner/jwala/ui/selenium/steps/configuration/ManageResourceRunSteps.java#L25) so it escaped me since I did not write it. I'll rename it though.

Sharvaribarve commented on jwala-integration-test/src/test/java/com/cerner/jwala/ui/selenium/steps/TearDownStep.java:
```diff
@@ -1,18 +1,120 @@
+            final RemoteSystemConnection remoteSystemConnection
+                    = new RemoteSystemConnection(sshUser, sshPwd, serverInfo.host, 22);
+            jschService.runShellCommand(remoteSystemConnection, "sc delete " + serverInfo.name, 300000);
+        }
```
Are we stopping the service before deleting it? are we planning on adding stop steps to  all cucumber files itself, as we are retrieving only stopped services?

jeddcuison replied:
For now no since the generate deploy that Rahul wrote that I refactored stops the service. If you are going to have a scenario that only does start, then we should add the stopping of the services in the cleanup. Since I'm merging your branch and if your branch just starts the services then I would have to put in the code.

jeddcuison committed:
Rename method names such that they reflect the Cucumber steps

jeddcuison committed:
Remove sleep

jeddcuison committed:
Adjust timeout and remove sleep

jeddcuison committed:
Use tomcat.media.remote.dir for Tomcat media and add JVM start steps

jeddcuison committed:
Adjust timeout

jeddcuison committed:
Add tomcat.media.remote.dir

jeddcuison committed:
Add JVM start and stop steps

jeddcuison committed:
Add required file

jeddcuison committed:
Add "web server" in the step statement to be more concise

jeddcuison committed:
Stop the service if running before deleting

In addition define repetitive variables as class variables

jeddcuison committed:
Use clickWhenReady since these steps can be used after expanding a row

Expanding a row might take a while and while doing so the buttons are not clickable hence usage of clickWhenReady

jeddcuison commented:
Please see #157
